source/particle/interpolator/cell_average.cc

```diff
@@ -0,0 +1,108 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/interpolator/cell_average.h>
+
+#include <deal.II/grid/grid_tools.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Interpolator
+    {
+      /**
+       * Return the cell-wise averaged properties of all tracers of the cell containing the
+       * given positions.
+       */
+      template <int dim>
+      std::vector<std::vector<double> >
+      CellAverage<dim>::properties_at_points(const std::multimap<types::LevelInd, Particle<dim> > &particles,
+                                             const std::vector<Point<dim> > &positions,
+                                             const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const
+      {
+        typename parallel::distributed::Triangulation<dim>::active_cell_iterator found_cell;
+
+        if (cell == typename parallel::distributed::Triangulation<dim>::active_cell_iterator())
+          {
+            // We can not simply use one of the points as input for find_active_cell_around_point
+            // because for vertices of mesh cells we might end up getting ghost_cells as return value
+            // instead of the local active cell. So make sure we are well in the inside of a cell.
+            Point<dim> approximated_cell_midpoint = positions[0];
+            if (positions.size() > 1)
+              {
+                Tensor<1,dim> direction_to_center;
+                for (unsigned int i = 1; i<positions.size()-1; ++i)
+                  direction_to_center += positions[i] - positions[0];
+                direction_to_center /= positions.size() - 1;
+                approximated_cell_midpoint += direction_to_center;
```

Comment:
is this algorithm a complicated way of computing (or approximating) the mean value of all of the positions? If so, why not just
```
  approximated_c_m = std::accumulate (positions.begin(), positions.end(), Point<dim>()) / positions.size();
```
?